geom/twkb_parser.go

```diff
@@ -0,0 +1,562 @@
+}
+
+func (p *TWKBParser) parseSize() error {
+	if len(p.twkb) <= p.pos {
+		return p.parserError("cannot unmarshal lacking size varint")
+	}
+	bytesRemaining, err := p.parseUnsignedVarint()
+	if err != nil {
+		return p.parserError("cannot unmarshal size varint malformed")
+	}
+	if uint64(p.pos)+bytesRemaining > uint64(len(p.twkb)) {
+		return p.parserError("cannot unmarshal remaining input smaller than size varint indicates")
+	}
+	return nil
+}
+
+func (p *TWKBParser) parseBBox() error {
+	// Parse the bounding box, but do not otherwise use it.
+	for i := 0; i < p.dimensions; i++ {
+		minVal, err := p.parseSignedVarint()
+		if err != nil {
+			return p.parserError("cannot unmarshal BBox min varint malformed")
+		}
+		p.bbox = append(p.bbox, minVal)
+
+		deltaVal, err := p.parseSignedVarint()
+		if err != nil {
+			return p.parserError("cannot unmarshal BBox delta varint malformed")
+		}
+		p.bbox = append(p.bbox, deltaVal)
+	}
+	return nil
+}
+
+func (p *TWKBParser) parsePoint() (Point, error) {
+	if p.isEmpty {
+		return NewEmptyPoint(p.ctype), nil
+	}
+	return p.nextPoint()
+}
+
+func (p *TWKBParser) nextPoint() (Point, error) {
+	var c Coordinates
+	c.Type = p.ctype
+
+	coords, err := p.parsePointArray(1)
+	if err != nil {
+		return Point{}, err
+	}
+
+	c.XY.X = coords[0]
+	c.XY.Y = coords[1]
+	if p.hasZ && p.hasM {
+		c.Z = coords[2]
+		c.M = coords[3]
+	} else if p.hasZ {
+		c.Z = coords[2]
+	} else if p.hasM {
+		c.M = coords[2]
+	}
+
+	pt, err := NewPoint(c, p.opts...)
+	return pt, p.annotateError(err)
+}
+
+func (p *TWKBParser) parseLineString() (LineString, error) {
+	if p.isEmpty {
+		return NewLineString(NewSequence(nil, p.ctype), p.opts...)
+	}
+	return p.nextLineString()
+}
+
+func (p *TWKBParser) nextLineString() (LineString, error) {
+	coords, _, err := p.parsePointCountAndArray()
+	if err != nil {
+		return LineString{}, err
+	}
+	ls, err := NewLineString(NewSequence(coords, p.ctype), p.opts...)
+	return ls, p.annotateError(err)
+}
+
+func (p *TWKBParser) parsePolygon() (Polygon, error) {
+	if p.isEmpty {
+		return NewPolygon(nil, p.opts...)
+	}
+	return p.nextPolygon()
+}
+
+func (p *TWKBParser) nextPolygon() (Polygon, error) {
+	numRings, err := p.parseUnsignedVarint()
+	if err != nil {
+		return Polygon{}, p.parserError("cannot unmarshal num rings varint malformed")
+	}
+
+	var rings []LineString
+	for r := uint64(0); r < numRings; r++ {
+		coords, numPoints, err := p.parsePointCountAndArray()
+		if err != nil {
+			return Polygon{}, err
+		}
+		if numPoints >= 2 {
+			// The spec says rings may need to be closed:
+			// "rings are assumed to be implicitly closed, so the first
+			//  and last point should not be the same"
+
+			// Note, some implementations can generate TWKB with the rings
+			// already closed. We wish to gracefully parse these cases too.
+			// So determine if the first and final coords differ.
+			finalPointDiffersFromFirst := false
+			for d := 0; d < p.dimensions; d++ {
+				first := coords[0+d]
```

Comment:
What's the reason for `0+d` instead of just `d`?

Reply:
It was just to reinforce the fact we're reading from the first point (at offset 0). It's obviously not strictly necessary. I can remove it.

Reply:
Okay, that makes sense, we can leave it there, or maybe put it in an `offset` variable if you feel it helps with readability. No strong opinion either way.